unitycatalog-ai/src/ucai/core/oss.py

```diff
@@ -0,0 +1,299 @@
+    """
+    Unity Catalog function calling client
+    """
+
+    def __init__(self, uc: Optional[Unitycatalog] = None, **kwargs: Any) -> None:
+        if uc is None:
+            uc = Unitycatalog()
+        if not isinstance(uc, Unitycatalog):
+            raise ValueError("client must be an instance of Unitycatalog")
+        self.uc = uc
+        # host all python functions within this cache
+        self.func_cache = {}
+        super().__init__()
+
+    @override
+    def create_function(
+        self,
+        *,
+        function_name: str,
+        routine_definition: str,
+        data_type: str,
+        parameters: Optional[List[Union[FunctionInfoInputParamsParameter, Dict[str, str]]]] = None,
+        timeout: Optional[float] = None,
+        replace: bool = False,
+        **kwargs,
+    ) -> FunctionInfo:
+        """
+        Create a function in Unity Catalog.
+
+        Args:
+            function_name: The full function name in the format <catalog_name>.<schema_name>.<function_name>.
+            routine_definition: The function definition.
+            data_type: The return data type. Allowed values are:
+                "BOOLEAN", "BYTE", "SHORT", "INT", "LONG", "FLOAT", "DOUBLE", "DATE", "TIMESTAMP",
+                "TIMESTAMP_NTZ", "STRING", "BINARY", "DECIMAL", "INTERVAL", "ARRAY", "STRUCT",
+                "MAP", "CHAR"
+            parameters: The input parameters.
+            timeout: The timeout in seconds.
+            replace: Whether to replace the function if it already exists. Default to False.
+        """
+        function_name = FullFunctionName.validate_full_function_name(function_name)
+        parameters = [validate_input_parameter(param) for param in parameters]
+        if data_type not in ALLOWED_DATA_TYPES:
+            raise ValueError(
+                f"Invalid data_type {data_type}, allowed values are {ALLOWED_DATA_TYPES}."
+            )
+        if replace and self.get_function(function_name, timeout=timeout):
+            _logger.info(f"Function {function_name} already exists, replacing it.")
+            self.delete_function(function_name, timeout=timeout)
+        function_info = CreateFunctionInfo(
+            catalog_name=function_name.catalog,
+            schema_name=function_name.schema,
+            name=function_name.function,
+            input_params=FunctionInfoInputParams(parameters=parameters),
+            data_type=data_type,
+            routine_body="EXTERNAL",
+            routine_definition=routine_definition,
+            security_type="DEFINER",
+            parameter_style="S",
+            external_language="PYTHON",
+            sql_data_access="NO_SQL",
+        )
+        return self.uc.functions.create(function_info=function_info, timeout=timeout, **kwargs)
+
+    @override
+    def create_python_function(
+        self, *, func: Callable[..., Any], catalog: str, schema: str, replace: bool = False
+    ) -> FunctionInfo:
+        raise NotImplementedError()
```

Comment:
I'll add in a follow-up :)

Reply:
https://github.com/unitycatalog/unitycatalog/pull/665 This is the PR

Reply:
perfect!